### PR TITLE
Resolves #1580: Lucene index maintainer should support delete where on group key

### LINKED ISSUE #1580: Lucene index maintainer should support delete where on group key
The Lucene index maintainer, if on a group key, should be able to support grouping key delete wheres. For example, if there's a grouping key on `concatenateFields("a", "b")`, it should support delete where for a delete where query bounded on either a predicate on `a` or on `a` and `b`.

This same kind of predicate is supported on the `TextIndexMaintainer` and the `RankIndexMaintainer`, so it should be straightforward to add.

### LINKED ISSUE #1597: Scans of empty Lucene index throws IndexNotFoundException
If a Lucene index is created but no data added, then queries from scanning the index throw `IndexNotFoundException`. Instead, this should return an empty cursor to indicate that there is no data

### PR DESCRIPTION
**Pull request by alecgrieser**
This adds support for `deleteRecordsWhere` to the `LuceneIndexMaintainer`. To do so, it needed to refactor a fair amount of the infrastructure around session management of `FDBDirectory`. This is because the `FDBDirectory` maintains a cache of the data inside of it, which means that if there's a range delete and then a read in the same transaction, the in-memory cache could result in deleted data being returned. The new code puts one object per index, the `FDBDirectoryManager`, in the transaction-level session cache, and then the `LuceneIndexMaintainer` uses it to grab a directory, or (in the case of a range delete) to invalidate directories in the cache. This also coalesced various commit checks to close these resources into a single commit check on the directory manager, which closes all outstanding objects.

This resolves #1580.

As part of the testing, I needed to be able to read data on an empty index, which through an error (see #1597). This adds logic to translate the error into an empty cursor, so this fixes #1597.

### CODE REVIEW

**Comment by alecgrieser** [comment]
This will conflict with #1584, but I've decoupled the two changes, so whichever merges first just needs to fix conflicts with the other one

**Comment by alecgrieser** [comment]
Okay, the latest push includes a fixed up merge. It also includes a fix for #1598, which was breaking the build. That's identical to the commit in #1599, which I broke out as it should be breaking everyone else, too, but if #1599 merges first, this PR may need a rebase

**Review by tian-yizuo — CHANGES_REQUESTED** [review]
LGTM basically, just a minor comment

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.lucene.index.MergeTrigger;
+import org.apache.lucene.index.TieredMergePolicy;
+import org.apache.lucene.search.suggest.analyzing.AnalyzingInfixSuggester;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+class FDBDirectoryWrapper implements AutoCloseable {
```
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryWrapper.java:49]: Added some to this class

**Comment by alecgrieser** [comment]
Rebase addresses merge conflicts. The primary one that was a little tricky to resolve was in `AutoCompleteSuggesterCommitCheckAsync.java` to use the `LuceneOptimizedWrappedBlendedInfixSuggester` when it creates the suggester

**Review by tian-yizuo — APPROVED** [review]

**Pull request merged**